pageserver/src/tenant/timeline/layer_manager.rs

```diff
@@ -21,94 +23,237 @@ use crate::{
+        let new_layer_map = f(LayerMap::clone(&*snapshot))?;
+        let old_layer_map = self.layer_map.swap(Arc::new(new_layer_map));
+        debug_assert_eq!(
+            Arc::as_ptr(&snapshot) as usize,
+            Arc::as_ptr(&old_layer_map) as usize,
+            "race detected when modifying layer map, use `update` instead of `initialize_update`."
+        );
+        Ok(())
+    }
+
+    /// Make an update to the layer map. Error will be returned only when the update function fails.
+    async fn update<T>(&self, f: impl FnOnce(LayerMap) -> Result<(LayerMap, T)>) -> Result<T> {
+        let _guard = self.state_lock.lock().await;
+        let snapshot = self.layer_map.load_full();
+        let (new_layer_map, data) = f(LayerMap::clone(&*snapshot))?;
+        let old_layer_map = self.layer_map.swap(Arc::new(new_layer_map));
+        debug_assert_eq!(
+            Arc::as_ptr(&snapshot) as usize,
+            Arc::as_ptr(&old_layer_map) as usize,
+            "race detected when modifying layer map, please check if `initialize_update` is used on the `update` code path."
+        );
+        Ok(data)
+    }
+}
+
+impl LayerSnapshot {
+    fn get_from_desc(&self, desc: &PersistentLayerDesc) -> Arc<dyn PersistentLayer> {
         self.layer_fmgr.get_from_desc(desc)
     }
 
     /// Get an immutable reference to the layer map.
     ///
-    /// We expect users only to be able to get an immutable layer map. If users want to make modifications,
-    /// they should use the below semantic APIs. This design makes us step closer to immutable storage state.
-    pub fn layer_map(&self) -> &LayerMap {
+    /// If a user needs to modify the layer map, they should get a write guard and use the semantic
+    /// functions.
+    fn layer_map(&self) -> &LayerMap {
         &self.layer_map
     }
 
-    /// Get a mutable reference to the layer map. This function will be removed once `flush_frozen_layer`
-    /// gets a refactor.
-    pub fn layer_map_mut(&mut self) -> &mut LayerMap {
-        &mut self.layer_map
+    fn contains(&self, layer: &Arc<dyn PersistentLayer>) -> bool {
+        self.layer_fmgr.contains(layer)
+    }
+}
+
+impl LayerManagerReadGuard {
+    pub fn get_from_desc(&self, desc: &PersistentLayerDesc) -> Arc<dyn PersistentLayer> {
+        self.snapshot.get_from_desc(desc)
+    }
+
+    /// Get an immutable reference to the layer map.
+    pub fn layer_map(&self) -> &LayerMap {
+        self.snapshot.layer_map()
+    }
+}
+
+impl LayerManagerWriteGuard {
+    /// Check if the layer file manager contains a layer. This should ONLY be used in the compaction
+    /// code path where we need to check if a layer already exists on disk. With the immutable layer
+    /// map design, it is possible that the layer map snapshot does not contain the layer, but the layer file
+    /// manager does. Therefore, use this function with caution.
+    pub(crate) fn contains(&self, layer: &Arc<dyn PersistentLayer>) -> bool {
```

Comment:
This could be `pub(super)` if it's only supposed to be used in compaction (which is the super module).